polyglot/piranha/visualize_rules_graph.py

```diff
@@ -0,0 +1,234 @@
+    """
+    Groups outgoing edges by rule/group.
+
+    All names are sanitized to replace empty spaces by `_`.
+    """
+    for edge_toml in edges_toml_dict['edges']:
+        from_node = sanitize_name(edge_toml['from'])
+        to_nodes: 'list[str]' = edge_toml['to']
+        scope = sanitize_name(edge_toml['scope'])
+        for to_node in to_nodes:
+            to_node = sanitize_name(to_node)
+            edge = Edge(to=to_node, scope=scope)
+            if from_node in outgoing_edges_by_node:
+                outgoing_edges_by_node[from_node].append(edge)
+            else:
+                outgoing_edges_by_node[from_node] = [edge]
+
+
+def group_rule_names_label(rule_names: 'list[str]') -> 'list[str]':
+    return [inline_name(rule_name) for rule_name in rule_names]
+
+
+def inline_name(rule_name: str) -> str:
+    """Should be called for rules under a group."""
+    if rule_name in cleanup_rules:
+        return f'{rule_name} (Cleanup Rule)'
+    else:
+        return rule_name
+
+
+def initialize_graph() -> graphviz.Digraph:
+    graph_attr = {
+        'label': str(args.title),
+        'labelloc': 't',
+        'fontsize': '30'
+    }
+    graph = graphviz.Digraph(filename=output_file_path, graph_attr=graph_attr)
+    graph.format = 'svg'
+    return graph
+
+
+def generate_graph_nodes():
+    generate_non_orphan_nodes()
+    generate_orphan_nodes()
+
+
+def generate_non_orphan_nodes():
+    for rule_name in all_non_orphan_rule_names():
+        if rule_name in rules_by_group_dict:
+            generate_node_for_group_rule(rule_name)
+        else:  # a rule not in a group
+            generate_node_for_non_group_rule(rule_name)
+
+        added_nodes.add(rule_name)
+
+
+def all_non_orphan_rule_names() -> 'list[str]':
+    node_names_with_only_outgoing_edges = [
+        node_name for node_name in rules_by_group_dict.keys() if node_name not in outgoing_edges_by_node.keys()
+    ]
+
+    non_orphan_rule_names = list(outgoing_edges_by_node.keys())
+    non_orphan_rule_names.extend(node_names_with_only_outgoing_edges)
+    non_orphan_rule_names.extend(cleanup_rules)
+    return non_orphan_rule_names
+
+
+def generate_node_for_group_rule(rule_name: str):
+    rule_names_in_group = rules_by_group_dict[rule_name]
+    for group_rule_name in rule_names_in_group:
+        added_nodes.add(group_rule_name)
+
+    rule_names_label: 'list[str]' = group_rule_names_label(
+        rule_names_in_group)
+    node_label = rule_name + '\\n\\n' + '\\n'.join(rule_names_label)
+
+    graph.node(rule_name, node_label, shape='record')
+
+
+def generate_node_for_non_group_rule(rule_name: str):
+    if rule_name not in added_nodes:
+        if rule_name in cleanup_rules:
+            node_label = f'{rule_name}\\n(Cleanup Rule)'
+            graph.node(rule_name, node_label)
+        else:
+            graph.node(rule_name)
+
+
+def generate_orphan_nodes():
+    # orphan_nodes don't have any group, not even 'Cleanup_Rule'
```

Comment:
Same comment as above. Are we using the term `orphan node` in a non standard way?